<html>
  <head>
    <meta charset="UTF-8" />
    <script src="modernizr-1.5.js"></script>
    <link href="lily_style.css" rel="stylesheet" type="text/css" />
  </head>
  <body>
    <header>
      <img src="logo.png">
      <h1>Lily</h1>
    </header>

    <aside>
      <ul>
        <li>Home</li>
        <li><a href="index.html">Introducing Lily</a></li>
        <li><a href="getting_started.html">Getting Started</a></li>
      </ul>
      <ul>
        <li>Core</li>
        <li><a href="syntax.html">Syntax</a></li>
        <li><a href="control_flow.html">Control flow</a></li>
        <li><a href="functions.html">Functions</a></li>
        <li><a href="builtin_types.html">Builtin types</a></li>
        <li><a href="generics.html">Generics</a></li>
        <li><a href="classes.html">Classes</a></li>
        <li><a href="enums.html">Enums</a></li>
        <li><a href="lambdas.html">Lambdas</a></li>
        <li><a href="error_handling.html">Error Handling</a></li>
        <li><a href="import.html">Importing</a></li>
      </ul>
      <ul>
        <li>Class Reference</li>
        <li><a href="boolean_type.html">boolean</a></li>
        <li><a href="integer_type.html">integer</a></li>
        <li><a href="bytestring_type.html">bytestring</a></li>
        <li><a href="string_type.html">string</a></li>
        <li><a href="function_type.html">function</a></li>
        <li><a href="list_type.html">list</a></li>
        <li><a href="hash_type.html">hash</a></li>
        <li><a href="tuple_type.html">tuple</a></li>
        <li><a href="any_type.html">any</a></li>
        <li><a href="exception_type.html">Exception</a></li>
      </ul>
      <ul>
        <li>Reference</li>
        <li><a href="sys_package.html">sys</a></li>
        <li><a href="server_package.html">server</a></li>
        <li><a href="postgre_package.html">postgre</a></li>
        <li><a href="builtins.html">builtins</a></li>
      </ul>
    </aside>
   <article>
     <section>
<h1>Error Handling</h1>
<p>Sometimes there are errors that occur from certain operations. A list has a subscript that goes too far, a division by zero, a bad format string, and more. When this occurs, Lily will raise an exception that can be caught.</p>
<p>To begin with, all exceptions that are defined within Lily have Exception as their base class. This class is defined internally as follows:</p>
<pre><code>class Exception(msg: string) {
    var @message = message
    var @traceback: list[string] = []
}
</code></pre>

<p>Since it is defined like a normal class, instead of having a representation baked into the interpreter, it is possible to subclass it to provide more specific exceptions. That is, in fact, how Lily provides the exception classes that are built into it. Those builtin classes that derive from Exception are as follows:</p>
<ul>
<li>
<p>DivisionByZeroError: Raised when there is a division or modulo by 0 or 0.0</p>
</li>
<li>
<p>IndexError: Raised when an list subscript is out of bounds.</p>
</li>
<li>
<p>BadTypecastError: Raised when a value is unable to cast to a certain type.</p>
</li>
<li>
<p>ValueError: Raised when a value is out of the expected range (ex: list::repeat with a negative index).</p>
</li>
<li>
<p>RuntimeError: Raised when there is an infinite loop in == or != (such as when there are recursive lists being compared).</p>
</li>
<li>
<p>KeyError: Raised when attempting to find a key in a hash that does not exist.</p>
</li>
<li>
<p>FormatError: Raised when string::format or printfmt receive the wrong number of args for their format specifiers (ex: <code>string::format("%s")</code>)</p>
</li>
<li>
<p>IOError: Raised when there is an issue calling a method of the file class (ex: Reading from a closed file, writing to a read-mode file, etc).</p>
</li>
<li>
<p>SyntaxError: This is an error raised internally by the interpreter when parsing syntax errors. It's not available publically.</p>
</li>
</ul>
<p>Instances of each of these errors (except SyntaxError) can be constructed using only a string as a message (ex: <code>ValueError::new("That value is incorrect.")</code>)</p>
<h1>Introducing try</h1>
<p>Lily provides a common try+catch-style block as a means of capturing errors. Here's a simple example:</p>
<pre><code>try:
    1 / 0
except DivisionByZeroError:
    print(&quot;Oh no, a division by zero.\n&quot;)
</code></pre>

<p>try blocks, like all blocks so far, comes in both a multi-line and a single-line version. In the above example, the error is caught and ignored. To catch the error and do something with it, an <code>as &lt;name&gt;</code> clause is required</p>
<pre><code>try:
    1 / 0
except DivisionByZeroError as e:
    print(e.message)
</code></pre>

<p>Perhaps the most important thing about exceptions is how they are caught. If a clause is found that is a base class of the exception being thrown, then that clause will be picked. This allows <code>except Exception</code> to catch all possible errors.</p>
<pre><code>try:
    1 / 0
except Exception as e:
    print(&quot;Caught it here.&quot;)
except DivisonByZeroError:
    print(&quot;Won't reach here.&quot;)
</code></pre>

<p>Not only will the clause for DivisionByZeroError not be reached (because Exception is a base class of the exception thrown). The following is trapped as a syntax error by Lily, because of that.</p>
<p>Errors can also be manually raised, using the <code>raise</code> keyword.</p>
<pre><code>try:
    raise ValueError::new(&quot;This value is wrong.&quot;)
except Exception:
    ...
</code></pre>

<p>An important thing to remember when handling exceptions is that each block is considered to be in a different scope. Because there is no guarantee that variables within the try clause cannot be used in an except clause</p>
<pre><code>try:
    var v = 1 / 0
except DivisionByZeroError:
    # This is a syntax error.
    printfmt(&quot;v is %d.\n&quot;, v)
</code></pre>

<h1>Custom Errors</h1>
<p>It's possible to create custom exceptions as follows:</p>
<pre><code>class MyError(message: string, code: integer) &gt; Exception(message)
{
    var @code = code
}

try:
    raise MyError::new(&quot;The code is too large&quot;, 900)
except MyError as e:
    printfmt(&quot;Error message: %s, code: %d.\n&quot;, e.message, e.code)
</code></pre>
     </section>
   </article>
     <footer>
      <p>
Lily is designed by Jesse Ray Adkins (<a href="https://github.com/jesserayadkins">Github</a>,
<a href="http://www.reddit.com/user/jesserayadkins">Reddit</a>).
      </p>
      <br />
      <p>
Logo source: <a href="http://sweetclipart.com/three-white-calla-lilies-1015">here</a>, via <a href="http://sweetclipart.com">sweetclipart.com</a>.
      </p>
    </footer>
</body>
</html>
